go/parquet/internal/encoding/byte_stream_split.go

```diff
@@ -0,0 +1,297 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encoding
+
+import (
+	"math"
+
+	"github.com/apache/arrow/go/v17/arrow"
+	"github.com/apache/arrow/go/v17/internal/bitutils"
+	"github.com/apache/arrow/go/v17/internal/utils"
+	"github.com/apache/arrow/go/v17/parquet"
+	"golang.org/x/xerrors"
+)
+
+type NumericByteStreamSplitType interface {
+	float32 | float64 | int32 | int64
+}
+
+type ByteStreamSplitType interface {
+	NumericByteStreamSplitType | parquet.FixedLenByteArray
+}
+
+// putByteStreamSplitNumeric is a generic implementation of the BYTE_STREAM_SPLIT encoding for fixed-width numeric types.
+func putByteStreamSplitNumeric[T NumericByteStreamSplitType](in []T, enc TypedEncoder, sink *PooledBufferWriter) {
+	numElements := len(in)
+	typeLen := enc.Type().ByteSize()
+	bytesNeeded := numElements * typeLen
+	sink.Reserve(bytesNeeded)
+
+	// Make sure len = cap so we can index into any loc rather than append
```

Comment:
should we make some assertion here?

Reply:
Sorry that comment may have been vague. We're actually doing that 2 lines below:
```
data = data[:cap(data)]
```

This ensures len = cap by reslicing `data`. I've updated the comments to clarify this.